### PR TITLE
[FIX] hr_expense: view receipts of expense line when clicked on

### PR DESCRIPTION
**Pull request by alialfie**
In an expense sheet, when an expense line is clicked on, its receipts are shown in the attachment viewer. This was broken in odoo/pull/142029 since it duplicates the attachments from the expense, leaving inconsistency between the attachment ID of the lines and the ones in the sheet. To fix this, the checksum is used instead to find which attachment in the sheet corresponds to the one in the line that was clicked on.

task-3758922

### CODE REVIEW

File: addons/hr_expense/static/src/js/tours/show_expense_receipt_tour.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+registry.category("web_tour.tours").add('show_expense_receipt_tour', {
+    test: true,
+    url: '/web',
+    steps: () => [
+        stepUtils.showAppsMenuItem(),
+       {
+            content: "Go to the Expenses app",
+            trigger: '.o_app[data-menu-xmlid="hr_expense.menu_hr_expense_root"]',
+        },
```
[the remark this answers is not in the document]
Comment (alialfie) [reply addons/hr_expense/static/src/js/tours/show_expense_receipt_tour.js:14]: Can you elaborate a bit more? Is using `goToAppSteps` better than how I did it? If so, why? I can see both being used in the codebase so not sure which way is better.

**Comment by alialfie** [comment]
Thanks for the review! Updated the PR @h4818

File: addons/hr_expense/views/hr_expense_views.xml [hunk 2]
```diff
@@ -886,6 +886,7 @@
                                     <field name="state" column_invisible="True"/>
                                     <field name="nb_attachment" column_invisible="True"/>
                                     <field name="message_main_attachment_id" column_invisible="True"/>
+                                    <field name="message_main_attachment_checksum" column_invisible="True"/>
```
[the remark this answers is not in the document]
Comment (h4818) [reply addons/hr_expense/views/hr_expense_views.xml:889]: I like this solution however... due to stable policy:
- we require an exception from runbot for the new field
- customers need to update, as you said

Perhaps we could try fetching a mapping of expense attachment ids to sheet attachment ids using an rpc in js?
or I'm open to any other ideas if you have one

**Review by h4818 — COMMENTED** [review]
Is there an alternative solution?

**Comment by alialfie** [comment]
@h4818 Ready for another review :) thanks!

File: addons/hr_expense/models/hr_expense_sheet.py [hunk 3]
```diff
@@ -756,3 +756,15 @@ def _get_expense_account_destination(self):
...
+    def get_expense_to_sheet_attachments_mapping(self):
+        """ Returns a mapping of expense attachment IDs to sheet attachment IDs"""
+        attachment_data = self.env['ir.attachment']._read_group([    # tuples (attachment, checksum)
+                ('checksum', 'in', self.attachment_ids.mapped('checksum')),
+                ('res_model', '=', 'hr.expense'),
+            ],
+            groupby=['id', 'checksum'],
+        )
+        sheet_checksums = {attachment.checksum: attachment.id for attachment in self.attachment_ids}
+        response = {group[0].id: sheet_checksums[group[1]] for group in attachment_data}
```
[the remark this answers is not in the document]
Comment (h4818) [reply addons/hr_expense/models/hr_expense_sheet.py:769]: EDIT: why do we need the read_group at all?
```py
    sheet_checksums = {attachment.checksum: attachment.id for attachment in self.attachment_ids}
    return {att.id: sheet_checksums[att.checksum] for att in self.expense_line_ids.attachment_ids}
```

**Review by h4818 — COMMENTED** [review]
@alialfie nice! this is safer for stable...
however its not working 100% - when adding a new expense line, the mapping needs to be retrieved again.

**Review by alialfie — COMMENTED** [review]
@h4818  Updated and pushed, thanks for the heads up!

**Review by h4818 — COMMENTED** [review]
almost there

**Comment by Xavier-Do** [comment]
Upgrade exception [#331](https://runbot.odoo.com/web/#id=331&view_type=form&model=runbot.upgrade.exception) added.

Waiting the forward-port of this pr, the exception will be applied on all builds, and create an inconsistent state for migrations.
**Please forward port this asap up to master without change.** Exception should be forward-ported before the end of the week.

PRs adding fields/models should be merged at the beginning of the week and all the forward ports should be merged in the same week.

If you need to apply any change before it reaches master, please notify runbot team.

Details:
```
field:hr.expense.message_main_attachment_checksum
```
cc @KangOl @nseinlet @aj-fuentes



As explained on discord i think it would be better to have another solution than needing to update the view, something like `fieldDependencies` even if i'm not verry familiar with this topic.  But I have no issue with the new field.

**Comment by h4818** [comment]
Thanks @Xavier-Do 
> i think it would be better to have another solution than needing to update the view, something like `fieldDependencies`
we've added `relatedFields` which means that the solution will work even without the view update. Since it's added defensively, it should be removed in master.

Considering the ci/template exception - could we remove it in master fw port or will a separate PR be better?

**Comment by h4818** [comment]
@robodoo r+

**Comment by h4818** [comment]
@robodoo rebase-ff

**Comment by Xavier-Do** [comment]
> Considering the ci/template exception - could we remove it in master fw port or will a separate PR be better?


I prefer that you do this in a separate pr, it is easier regarding upgrade scripts, testing,.... also, it is a good idea to wait a little and check that we removed the exceptions before working on your upgrade script to ensure that the exception does not mess with the change.

**Pull request closed without merging**